boa_engine/src/vm/mod.rs

```diff
@@ -2270,6 +2277,106 @@ impl Context {
+                let promise = Promise::promise_resolve(
+                    self.intrinsics().constructors().promise().constructor(),
+                    value,
+                    self,
+                )?;
+
+                // 3. Let fulfilledClosure be a new Abstract Closure with parameters (value) that captures asyncContext and performs the following steps when called:
+                // 4. Let onFulfilled be CreateBuiltinFunction(fulfilledClosure, 1, "", « »).
+                let on_fulfilled = FunctionBuilder::closure_with_captures(
+                    self,
+                    |_this, args, captures, context| {
+                        // a. Let prevContext be the running execution context.
+                        // b. Suspend prevContext.
+                        // c. Push asyncContext onto the execution context stack; asyncContext is now the running execution context.
+                        // d. Resume the suspended evaluation of asyncContext using NormalCompletion(value) as the result of the operation that suspended it.
+                        // e. Assert: When we reach this step, asyncContext has already been removed from the execution context stack and prevContext is the currently running execution context.
+                        // f. Return undefined.
+
+                        std::mem::swap(&mut context.realm.environments, &mut captures.0);
+                        std::mem::swap(&mut context.vm.stack, &mut captures.1);
+                        context.vm.push_frame(captures.2.clone());
+
+                        context.vm.frame_mut().generator_resume_kind = GeneratorResumeKind::Normal;
+                        context.vm.push(args.get_or_undefined(0));
+                        context.run()?;
+
+                        captures.2 = *context
+                            .vm
+                            .pop_frame()
+                            .expect("generator call frame must exist");
+                        std::mem::swap(&mut context.realm.environments, &mut captures.0);
+                        std::mem::swap(&mut context.vm.stack, &mut captures.1);
+
+                        Ok(JsValue::undefined())
+                    },
+                    (
+                        self.realm.environments.clone(),
+                        self.vm.stack.clone(),
+                        self.vm.frame().clone(),
+                    ),
+                )
+                .name("")
+                .length(1)
+                .build();
+
+                // 5. Let rejectedClosure be a new Abstract Closure with parameters (reason) that captures asyncContext and performs the following steps when called:
+                // 6. Let onRejected be CreateBuiltinFunction(rejectedClosure, 1, "", « »).
+                let on_rejected = FunctionBuilder::closure_with_captures(
+                    self,
+                    |_this, args, captures, context| {
```

Comment:
```suggestion
                    |_this, args, (environment, value, frame), context| {
```

I think it would be more readable if we use argument destructuring here, instead of accessing tuples through index.

Reply:
Good call. Should be fixed now.